### PR TITLE
[RLlib] Policy V2 unify APIs (part 1)

### PR DESCRIPTION
**Pull request by sven1977**
This PR addresses the need to unify the different new Policy classes: TorchPolicyV2, EagerTFPolicyV2, and DynamicTFPolicyV2 to (mostly) use the same top-level API (inherited from Policy). Moving toward Ray 2.0, we should tie up this project in a timely fashion via these PRs here. A few more cleanup PRs will follow this one. The changes in here are:
* Establish `compute_actions` as the only way to customize action computations (including exploration!).
* Retire `extra_action_out_(fn)?`. In order to add extra action out fetches or computations to each action computing forward pass, simply override `compute_actions`, call super() therein (to get the "normal" behavior), and then add extra data to the returned extra_out dict. Then return that. See PPO's value-function handling (in this very PR) as an example on how to do this.
* Retire `action_sampler_fn` and `action_distribution_fn` (see above: Use `compute_actions()` instead).
* Soft-deprecate `compute_actions_from_input_dict()` and `compute_single_action()`. Again - only `compute_actions` should be used from here on.
* Establish Policy sub-classing as the only way to customize a policy. Deprecate `build_tf_policy` and `build_policy_class` utility methods.
* Establish `make_model` as the only way to custom-create models. Retire the torch-specific `make_model_and_action_dist`. Custom action distributions should either be specified via the `config.model.custom_action_dist` key OR inside an overridden `compute_actions` method.
* Establish `make_optimizer` as the only way to custom-create local optimizers for a policy. Retire/soft-deprecate the existing `optimizer()` methods. Their naming is confusing and hints toward a property (self.optimizer, similar to self.model), which they are not.
* Establish `stats_fn()` as the only way to create custom stats after the loss(es) have been calculated.
* Rename `maybe_initialize_optimizer_and_loss` into `initialize_optimizer_and_loss_if_necessary` to abide to RLlib-wide naming convention (`..._if_necessary()`).

TODO: In a follow up PR, we need to unify and cleanup all gradient-computing/stats related methods in TF- and TorchPolicies. There currently exists a zoo of different names and methods, which leads to a ton of confusion:
* compute_gradients, compute_gradients_fn, apply_gradients_fn, and apply_gradients
* TF-world: `grad_stats_fn()`, `extra_learn_fetches_fn()`, `extra_compute_grad_fetches()`, `extra_compute_grad_feed_dict()`, `gradients()`, `build_apply_op()`.
* TorchWorld: `extra_grad_process()`, `extra_compute_grad_fetches`(), 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/core-concepts.rst [hunk 1]
```diff
@@ -77,26 +77,72 @@ each controlling one or more agents:
 
 .. image:: images/multi-flat.svg
 
-Policies can be implemented using `any framework <https://github.com/ray-project/ray/blob/master/rllib/policy/policy.py>`__.
-However, for TensorFlow and PyTorch, RLlib has
-`build_tf_policy <rllib-concepts.html#building-policies-in-tensorflow>`__ and
-`build_torch_policy <rllib-concepts.html#building-policies-in-pytorch>`__ helper functions that let you
-define a trainable policy with a functional-style API, for example:
+Custom policies can be implemented using RLlib's :py:class:`~ray.rllib.policy.policy.Policy` class.
```
Comment (sven1977) [inline doc/source/rllib/core-concepts.rst:80]: Documentation changes are still in progress. Please ignore for now ...

File: rllib/agents/a3c/a3c_torch_policy.py [hunk 2]
```diff
@@ -116,13 +116,6 @@ def loss(
 
         return total_loss
 
-    @override(TorchPolicyV2)
```
Comment (sven1977) [inline rllib/agents/a3c/a3c_torch_policy.py:119]: This is not needed. It does exactly the same thing as the default implementation.

File: rllib/agents/impala/vtrace_tf_policy.py [hunk 3]
```diff
@@ -219,7 +219,7 @@ def __init__(self):
 
     # TODO: maybe standardize this function, so the choice of optimizers are more
     # predictable for common agents.
-    def optimizer(
+    def make_optimizer(
```
Comment (sven1977) [inline rllib/agents/impala/vtrace_tf_policy.py:222]: Make_model, make_optimizer.
Note that `optimizer()` still works, but should be deprecated.

File: rllib/agents/ppo/appo_tf_policy.py [hunk 4]
```diff
@@ -182,6 +182,27 @@ def __init__(
         def make_model(self) -> ModelV2:
             return make_appo_model(self)
 
+        @override(base)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/appo_tf_policy.py:185]: `extra_action_fetches/outs` should be retired for keeping the API simple and expressive.

File: rllib/agents/ppo/tests/test_appo.py [hunk 5]
```diff
@@ -5,7 +5,7 @@
 from ray.rllib.policy.sample_batch import DEFAULT_POLICY_ID
 from ray.rllib.utils.metrics.learner_info import LEARNER_INFO, LEARNER_STATS_KEY
 from ray.rllib.utils.test_utils import (
-    check_compute_single_action,
+    check_compute_actions_v2,
```
Comment (sven1977) [inline rllib/agents/ppo/tests/test_appo.py:8]: Simpler version for PolicyV2s

File: rllib/evaluation/sampler.py [hunk 6]
```diff
@@ -1172,11 +1175,26 @@ def _do_policy_eval(
             policy: Policy = _get_or_raise(policies, policy_id)
 
         input_dict = sample_collector.get_inference_input_dict(policy_id)
-        eval_results[policy_id] = policy.compute_actions_from_input_dict(
-            input_dict,
-            timestep=policy.global_timestep,
-            episodes=[active_episodes[t.env_id] for t in eval_data],
-        )
+        # New Policy API: Only method to ever use (and override) to get and compute
```
Comment (sven1977) [inline rllib/evaluation/sampler.py:1178]: Sampler has to distinguish between `compute_actions` (PolicyV2s) and `compute_actions_from_input_dict` (old Policy classes).

File: rllib/agents/impala/vtrace_tf_policy.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/agents/impala/vtrace_tf_policy.py:222]: Do I understand correctly that with this we could infuse our own optimizers into the RL algorithms and `Adam` is not anymore the single way to go? That would be great for research to replicate papers that use specific algorithms

File: rllib/agents/impala/vtrace_torch_policy.py [hunk 7]
```diff
@@ -170,7 +170,7 @@ def __init__(self):
         pass
 
     @override(TorchPolicyV2)
-    def optimizer(
+    def make_optimizer(
         self,
     ) -> Union[List["torch.optim.Optimizer"], "torch.optim.Optimizer"]:
         if self.config["opt_type"] == "adam":
```
Comment (simonsays1980) [inline rllib/agents/impala/vtrace_torch_policy.py:176]: Here, it looks like that a custom optimizer is brought in via the configuration parameter `opt_type`

File: rllib/agents/ppo/appo_tf_policy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/agents/ppo/appo_tf_policy.py:185]: I Appreciate a lot the cleaning of the `Policy` from `compute_single_action` and `compute_action_from_input_dict`, as the three functions together were irritating at first. 

So `extra_outs` replace the agent's `info` dict? How is it then with the `Output API`? Are these `extra_outs` also written (like the former `info`) into the output JSONs? 

Also, how is it with the `input_dict`? Using `compute_actions` in a custom loop, could the user just write `input_dict={obs: observation_array}`?

**Comment by simonsays1980** [comment]
This looks very nice (and like a huge bunch of work)! Great work @sven1977 ! 

Some notes:

- I appreciate a lot the move from functional towards more OOP-style. This makes in my opinion a lot of things clearer and for custom policies we had to subclass anyway.
- Regarding naming of the TF policies, I like the names, but I guess they could appear to users with less rllib history unintuitive. So maybe `TF1` and `TF2` are more intuitive for them.
- Towards tthe `compute_action()` function: as written in the comment above, I think it is important that the `extra_outs` also make it automatically into the output files of the `Output API`, but I guess this is already working.
    - For the `input_dict` to be used easily, users should have the option to pass in a simple dictionary like `{obs: user_observation_array}` so they can use it for evaluation, deployment, or custom loops without much hustle (I guess this is possible, but I had to jump over the code due to some deadlines).
- In regard to the `make_optimizer` function and my two comments above: I consider custom optimizers (even if only from Keras/Torch) as a very useful feature as changing optimizers can change in some situations training a lot. I have not yet figured out, if this feature comes with `make_optimizer`, but it looks like :) 
- Reading through some code I really appreciate that a lot more useful comments come with it - this is very helpful for contributing. Thumbs up! 
- The naming Odyssee that will still come in the follow-up PRs is advantageous for users as well as us contributors as often names were not that intuitive and functions did sometimes way more than what their names promised. In regard to this renaming will help to make the code more readable (as you know I am always for readability :)) 

Thanks for a lot of great work!
